File: typescript/test/deserialize.test.ts
import { describe, it, expect } from "vitest";
import { hasJsonTag } from "../src/index.js";
import type { Shape, Vehicle } from "./testTypes.js";

// Interop JSON — identical strings used across all 12 language libraries
const CIRCLE_JSON = '{"#type":"Circle","radius":4}';
const RECTANGLE_JSON = '{"#type":"Rectangle","width":10,"height":5}';
const CAR_JSON = '{"#type":"Car","make":"Toyota"}';
const TRUCK_JSON = '{"#type":"Truck","payloadTons":5}';

describe("JSON.parse", () => {
  it("when deserializing Circle, resolves correct JsonTagged type", () => {
    // Act
    const shape = JSON.parse(CIRCLE_JSON) as Shape;

    // Assert
    expect(hasJsonTag(shape, "Circle")).toBe(true);
    if (hasJsonTag(shape, "Circle")) {
      expect(shape.radius).toBe(4);
    } else {
      expect.unreachable("expected Circle");
    }
  });

  it("when deserializing Rectangle, resolves correct JsonTagged type", () => {
    // Act
    const shape = JSON.parse(RECTANGLE_JSON) as Shape;

    // Assert
    expect(hasJsonTag(shape, "Rectangle")).toBe(true);
    if (hasJsonTag(shape, "Rectangle")) {
      expect(shape.width).toBe(10);
      expect(shape.height).toBe(5);
    } else {
      expect.unreachable("expected Rectangle");
    }
  });

  it("when deserializing Car, resolves correct JsonTagged type", () => {
    // Act
    const vehicle = JSON.parse(CAR_JSON) as Vehicle;

    // Assert
    expect(hasJsonTag(vehicle, "Car")).toBe(true);
    if (hasJsonTag(vehicle, "Car")) {
      expect(vehicle.make).toBe("Toyota");
    } else {
      expect.unreachable("expected Car");
    }
  });

  it("when deserializing Truck, resolves correct JsonTagged type", () => {
    // Act
    const vehicle = JSON.parse(TRUCK_JSON) as Vehicle;

    // Assert
    expect(hasJsonTag(vehicle, "Truck")).toBe(true);
    if (hasJsonTag(vehicle, "Truck")) {
      expect(vehicle.payloadTons).toBe(5);
    } else {
      expect.unreachable("expected Truck");
    }
  });

  it("when #type is not first property, still resolves correct JsonTagged type", () => {
    // Act
    const shape = JSON.parse('{"radius":4,"#type":"Circle"}') as Shape;

    // Assert
    expect(hasJsonTag(shape, "Circle")).toBe(true);
    if (hasJsonTag(shape, "Circle")) {
      expect(shape.radius).toBe(4);
    } else {
      expect.unreachable("expected Circle");
    }
  });
});

File: ocaml/test/deserialize_test.ml
(* Interop JSON — identical strings used across all 12 language libraries *)
let circle_json = {|{"#type":"Circle","radius":4}|}
let rectangle_json = {|{"#type":"Rectangle","width":10,"height":5}|}
let car_json = {|{"#type":"Car","make":"Toyota"}|}
let truck_json = {|{"#type":"Truck","payloadTons":5}|}

let tests =
  [
    Alcotest.test_case
      "decode, when deserializing circle, resolves correct type" `Quick
      (fun () ->
        (* Act *)
        let result =
          Yojson.Safe.from_string circle_json
          |> Test_types.decode_shape Json_tag.decode
        in
        (* Assert *)
        match result with
        | Ok (Test_types.Circle { radius = 4.0 }) -> ()
        | _ -> Alcotest.fail "expected Circle with radius 4.0");
    Alcotest.test_case
      "decode, when deserializing rectangle, resolves correct type" `Quick
      (fun () ->
        (* Act *)
        let result =
          Yojson.Safe.from_string rectangle_json
          |> Test_types.decode_shape Json_tag.decode
        in
        (* Assert *)
        match result with
        | Ok (Test_types.Rectangle { width = 10.0; height = 5.0 }) -> ()
        | _ -> Alcotest.fail "expected Rectangle with width 10.0 and height 5.0");
    Alcotest.test_case
      "decode, when deserializing car, resolves correct type" `Quick
      (fun () ->
        (* Act *)
        let result =
          Yojson.Safe.from_string car_json
          |> Test_types.decode_vehicle Json_tag.decode
        in
        (* Assert *)
        match result with
        | Ok (Test_types.PersonalVehicle { make = "Toyota" }) -> ()
        | _ -> Alcotest.fail "expected PersonalVehicle with make Toyota");
    Alcotest.test_case
      "decode, when deserializing truck, resolves correct type" `Quick
      (fun () ->
        (* Act *)
        let result =
          Yojson.Safe.from_string truck_json
          |> Test_types.decode_vehicle Json_tag.decode
        in
        (* Assert *)
        match result with
        | Ok (Test_types.TransportationVehicle { payload_tons = 5.0 }) -> ()
        | _ -> Alcotest.fail "expected TransportationVehicle with payloadTons 5.0");
    Alcotest.test_case
      "decode, when hash type is not first property, still deserializes" `Quick
      (fun () ->
        (* Act *)
        let result =
          Yojson.Safe.from_string {|{"radius":4,"#type":"Circle"}|}
          |> Test_types.decode_shape Json_tag.decode
        in
        (* Assert *)
        match result with
        | Ok (Test_types.Circle { radius = 4.0 }) -> ()
        | _ -> Alcotest.fail "expected Circle with radius 4.0");
    Alcotest.test_case
      "decode, when missing hash type, returns error" `Quick
      (fun () ->
        (* Act *)
        let result =
          Yojson.Safe.from_string {|{"radius":4}|}
          |> Test_types.decode_shape Json_tag.decode
        in
        (* Assert *)
        match result with
        | Error _ -> ()
        | Ok _ -> Alcotest.fail "expected error for missing #type");
  ]
